hedera-node/src/main/java/com/hedera/services/ledger/interceptors/EndOfStakingPeriodCalculator.java

```diff
@@ -0,0 +1,149 @@
+		this.merkleNetworkContextSupplier = merkleNetworkContextSupplier;
+		this.syntheticTxnFactory = syntheticTxnFactory;
+		this.recordsHistorian = recordsHistorian;
+		this.creator = creator;
+		this.properties = properties;
+	}
+
+	public void updateNodes(final Instant consensusTime) {
+		final var stakingInfo = stakingInfoSupplier.get();
+		final var merkleNetworkContext = merkleNetworkContextSupplier.get();
+
+		// skip end of staking period calculations if the rewards are not yet activated.
+		if (!merkleNetworkContext.areRewardsActivated()) {
+			return;
+		}
+
+		final var rewardRate = Math.min(stakingRewardsAccountBalance(),
+				properties.getLongProperty("staking.rewardRate"));
+
+		long updatedTotalStakedRewardStart = 0L;
+		long updatedTotalStakedStart = 0L;
+		List<NodeStake> nodeStakingInfos = new ArrayList<>();
+
+		for (final var nodeNum : stakingInfo.keySet()) {
+			final var merkleStakingInfo = stakingInfo.getForModify(nodeNum);
+			merkleStakingInfo.updateRewardSumHistory(rewardRate, merkleNetworkContext.getTotalStakedRewardStart());
+
+			final var totalStake = merkleStakingInfo.getStakeToReward() + merkleStakingInfo.getStakeToNotReward();
+			final var minStake = merkleStakingInfo.getMinStake();
+			final var maxStake = merkleStakingInfo.getMaxStake();
+
+			if (totalStake > maxStake) {
+				merkleStakingInfo.setStake(maxStake);
+			} else if (totalStake < minStake) {
+				merkleStakingInfo.setStake(0);
+			} else {
+				merkleStakingInfo.setStake(totalStake);
+			}
+
+			final var stakedRewardUsed = Math.min(merkleStakingInfo.getStakeToReward(), merkleStakingInfo.getStake());
+			updatedTotalStakedRewardStart += stakedRewardUsed;
+			updatedTotalStakedStart += merkleStakingInfo.getStake();
+
+			nodeStakingInfos.add(NodeStake.newBuilder()
+					.setStake(merkleStakingInfo.getStake())
+					.setStakeRewarded(merkleStakingInfo.getStakeToReward())
+					.build());
+		}
+		merkleNetworkContext.setTotalStakedRewardStart(updatedTotalStakedRewardStart);
+		merkleNetworkContext.setTotalStakedStart(updatedTotalStakedStart);
+
+		// create a synthetic txn with this computed data
+		final var syntheticNodeStakeUpdateTxn =
+				syntheticTxnFactory.nodeStakeUpdate(
+						getMidnightTime(consensusTime),
+						rewardRate,
+						nodeStakingInfos);
+
+		recordsHistorian.trackFollowingChildRecord(
+				DEFAULT_SOURCE_ID, syntheticNodeStakeUpdateTxn,
+				creator.createSuccessfulSyntheticRecord(
+						NO_CUSTOM_FEES,
+						new SideEffectsTracker(),
+						END_OF_STAKING_PERIOD_CALCULATIONS_MEMO));
+	}
+
+	private Timestamp getMidnightTime(Instant consensusTime) {
+		final var secsInDay = 86400L;
+		final var originalSecs = consensusTime.getEpochSecond();
+		final long midNightSecs = (originalSecs / secsInDay) * secsInDay;
+
+		return Timestamp.newBuilder()
+				.setSeconds(midNightSecs)
+				.build();
+	}
+	private long stakingRewardsAccountBalance() {
+		return accounts.get().get(EntityNum.fromInt(800)).getBalance();
```

Comment:
We're repeating this constant in enough places, I think it deserves a field [in `AccountNumbers`](https://github.com/hashgraph/hedera-services/blob/master/hedera-node/src/main/java/com/hedera/services/config/AccountNumbers.java#L32) just to improve readability.

Low priority though, could be a touch-up PR later.